### PR TITLE
Adding HTTP /terminate path to Dikastes for graceful termination

### PR DESCRIPTION
**Pull request by Josh-Tigera**
## Description
Introduces an HTTP path "/terminate" for graceful termination of Dikastes to allow for successful completion of Kubernetes jobs. Envoy has an analogous /quitquitquit path so this Dikastes equivalent should be familiar with anyone who has used Envoy outside of Calico

Tested via newly introduced unit tests as well as by manually creating a Job that includes a POST operation via curl to shutdown both Dikastes and Envoy and observing that the Job is successfully marked as completed.

## Related issues/PRs
fixes [5227](https://github.com/projectcalico/calico/issues/5227)

## Todos

- [x] Tests
- [x] Documentation
- [x] Release note

## Release Note

```release-note
Add HTTP /terminate endpoint for graceful termination of Dikastes sidecar container to facilitate Kubernetes Job completion
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: calico/reference/architecture/overview.md [hunk 1]
```diff
@@ -86,6 +86,8 @@ Confd dynamically generates BIRD configuration files based on the updates to dat
 
 (Optional) {{site.prodname}} enforces network policy for workloads at both the Linux kernel (using iptables, L3-L4), and at L3-L7 using a Envoy sidecar proxy called Dikastes, with cryptographic authentication of requests. Using multiple enforcement points establishes the identity of the remote endpoint based on multiple criteria. The host Linux kernel enforcement protects your workloads even if the workload pod is compromised, and the Envoy proxy is bypassed.
 
+> **Note**: Dikastes can be terminated by issuing an HTTP POST request to /terminate on localhost at the HTTP port specified with the --http-port option when starting Dikastes (default 7777). This is to allow for graceful termination so that Kubernetes Jobs can complete successfully and is analogous to Envoy's /quitquitquit. eg. `curl -XPOST http://127.0.0.1:7777/terminate`
```
Comment (caseydavenport) [inline calico/reference/architecture/overview.md:89]: We usually put CLI options surrounded by backticks I think, so `--http-port`

**Review by caseydavenport — CHANGES_REQUESTED** [review]

**Review by caseydavenport — COMMENTED** [review]
Couple of minor comments but looking good.

File: app-policy/cmd/dikastes/dikastes.go [hunk 2]
```diff
@@ -172,3 +203,34 @@ func runClient(arguments map[string]interface{}) {
...
+func (h *httpTerminationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.termChan <- true
+	if _, err := io.WriteString(w, "terminating Dikastes\n"); err != nil {
+		log.Fatalf("error writing HTTP response: %v", err)
+	}
+}
+
+func (h *httpTerminationHandler) RunHTTPServer(arguments map[string]interface{}) (*http.Server, *sync.WaitGroup) {
+	httpServerPort := arguments["--http-port"].(string)
+	httpServerPortStr := fmt.Sprintf(":%s", httpServerPort)
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply app-policy/cmd/dikastes/dikastes.go:220]: Yep, if you don't specify an address it will bind to all available addresses

**Comment by brentmjohnson** [comment]
Any interest in releasing this one into the v3.23?  Ran into some scenarios where this would be really valuable (for 3rd party operator managed job manifests).

**Comment by Josh-Tigera** [comment]
Hi @brentmjohnson,

We don't typically port new content back into older releases, we restrict that to bug fixes only so if this functionality is of use to you please consider upgrading once v3.24 has been released.

**Pull request merged**